### DISCUSSION TITLE
Compose definitions

### DISCUSSION THREAD

**Discussion by lax4mike** [ramda/types#31] [General]
Hello!  

I was messing around with compose and discovered something interesting.  The type definitions for `compose` are almost identical to `pipe` (only the function order is reversed), but it doesn't seem to behave the same. Consider [these compositions](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAKjgQwM5wEpwGZQiOAIimRABNlCBuAKBuwFcA7AYxmAibhf0lQFMATAB4MABgA0mAIxSMggHwAKGnByCAXHCXItGaQEo4AXgWZBE1TmladesUdMzLB27swOTZ+XADeVqH4YBiguJQAPe0czbEElbGkIgwNaAF86HiZUeB5wCAEAFX5sk25efKEVNQwAOjJ+bGQGABsYAoglAHJG5uaAI2QWAGtOg0tqmqYYAAslQzgAegW4YWE4GABPMGKWKGAweGB0ZGbA5DINsqZGATIaAyUAbVk4QQBdFLpGVnZOODB9kJRJJnOZlFYEm4ol5nBDNNp3PpouYXFCPMifP41IFgqFtJF0TDYvFEuFkmk6EsylBAmxmpcmAwQH1+FA4AAfODZPZMADmNEyJQB2yKJWM-0BcSstSms0M40wdQaTVa7S6PX6gxGY3uTxe7xSQA), one with `compose` and one with `pipe`:
```typescript
const poo = R.compose(
  R.defaultTo('fallback'),
  R.nth(1)
)([1, 2]);

const poo2 = R.pipe(
  R.nth(1),
  R.defaultTo('fallback')
)([1, 2]);
```
`pipe` is fine with this, and `poo2` is correctly `string | number`.  On the other hand, `compose` is getting confused and says `poo` is `unknown`.  Additionally, there's a big red squiggle under `R.nth(1)` complaining about the `unknown`.  

Given that the type definitions are almost identical, I'm not sure why `compose` wouldn't work the same as `pipe`.  Any idea what's going on here?

**Comment by Harris-Miller** [comment]
That is very odd. Stranger still that if you put a `R.toString` at the "end" of either, they both then work as expected

[See here](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAKjgQwM5wEpwGZQiOAcimRABNlCBuAWACh6BjCAO1XmxYGE4BeTAHTNwEVAFMAFPTiCYEAMowowFgHMANNMFkx2ZAFcANjAAqECYT2HDAI2SMA1oQCUmujIwCWMABYSAjM70zrTuMpxc9AD0UXAAegD89Eys7DjcUHzpXBIA2v7qcABMALoh0bEy8UkMdMxsHCwAClmeYMBgklqe3n6Bbh4COnpGpuaWyNZ2ji4DsgpKKqrBoVWcTRXVyXWpjU2Z-Ot5BcVloTFVW3RAA)

**Reply by lax4mike** [reply]
Adding `toString` at the end makes sense to me because the signature is `unknown -> string` so it would return `string` no matter what, essentially masking the `unknown` error that `compose` introduced.

**Comment by lax4mike** [comment]
I'm reaching out for some [help on discord](https://discord.com/channels/508357248330760243/1081578839915905114/1105120145166319689).  It looks like someone else had this question before and there was no resolution 🫤

**Comment by lax4mike** [comment]
I also posted [on stack overflow](https://stackoverflow.com/questions/76266108/type-definitions-for-ramda-compose-function), hopefully someone can help us!

**Comment by lax4mike** [comment]
I also cross posted to [ramda/ramda](https://github.com/ramda/ramda/issues/3383)
